### PR TITLE
[MRG] add manifests to support fast `Index.select(...)` and lazy loading

### PR DESCRIPTION
**Pull request by ctb**
This PR adds manifests to sourmash, in order to support rapid `Index.select(...)` (including with picklists) on very large databases. The core idea is that manifests contain the signature _metadata_, including (potentially) the storage location for the raw signature, and can be repeatedly selected upon before actually loading the data.

In particular, this enables cool stuff like rapid extraction of signatures from arbitrarily large collections per https://github.com/dib-lab/sourmash/issues/1365.

This PR:
- creates a `CollectionManifest` class that stores manifest "rows" containing signature metadata, and can be serialized easily (currently to CSV);
- extends `SignaturePicklist` to do matching on manifest metadata rather than on signature objects only, via `matches_manifest_row(...)`;
- refactors the `MultiIndex` class into an in-memory class that uses manifests.
- provides a `sourmash sig manifest` function to create manifests.
- provides manifest functionality for `SBT` (via https://github.com/sourmash-bio/sourmash/pull/1597) and `ZipFileLinearCollection` on both write and read.

One important part of this PR is the addition of `Index._signatures_with_internal()` which must be defined by `Index` classes that support manifest _creation_; it yields internal locations for signatures in a way that is not currently supported by any other mechanism. This is something that we'll need to revisit in the future, and perhaps refactor into `Storage` classes.

TODO:
- [x] revisit `signatures_with_internal`
- [x] write tests for `ZipFileLinearIndex` with and without manifests
- [x] write tests for manifest generation on a variety of saved indices

## a demonstration

Below, we run a prefetch of a signature against 48,000 signatures in a zipfile collection, which yields 13 matches (for this query). We then use the picklist functionality in `sourmash sig extract` with the `match_md5` column from the prefetch results to  extract just the relevant signatures.

With this PR, the `sourmash sig extract` takes about 2 seconds. With #1588 (which supports picklists but not any special zipfile interaction), it takes a few minutes.

```
% sourmash prefetch podar-ref/63.fa.sig gtdb-r202.genomic-reps.k31.zip -o 63.prefetch.csv
...
(takes a few minutes, yields a prefetch.csv with 13 results)
...
% sourmash sig extract --picklist 63.prefetch.csv:match_md5:md5prefix8 \
         gtdb-r202.genomic-reps.k31.zip -o /tmp/abc.zip
picking column 'match_md5' of type 'md5prefix8' from '63.prefetch.csv'
loaded 13 distinct values into picklist.
found manifest when loading gtdb-r202.genomic-reps.k31.zip
found manifest when loading gtdb-r202.genomic-reps.k31.zip
.signatures() found manifest!
loaded 13 sigs from 'gtdb-r202.genomic-reps.k31.zip'
loaded 13 total that matched ksize & molecule type
extracted 13 signatures from 1 file(s)
for given picklist, found 13 matches of 13 total
```

### CODE REVIEW

**Comment by ctb** [comment]
with `LazyMultiIndex`, we can now support fast lazy selection on collections of `Index` objects. The key innovation here is to maintain the manifests separately from the index objects, and do selects on the manifests; and then use the manifests as picklists for `Index.select(...)` when the time comes to actually do the loading.

Then, as the underlying `Index` objects support lazy loading, so does this code 😁 .

## example

if `lazy-mi2.pathlist` contains:
```
gtdb-r202.genomic-reps.k31.zip
./podar-ref/
./podar-ref/60.fa.sig
```

then the following code yields the relevant signatures:

```
import sourmash
import sourmash
from sourmash.index import LazyMultiIndex
from sourmash.sig.picklist import SignaturePicklist

print('creating lazy-mi:')
mi = LazyMultiIndex.load_from_pathlist('lazy-mi2.pathlist')

from sourmash.sig.picklist import SignaturePicklist

picklist = SignaturePicklist(None, None, 'md5prefix8')
picklist.init(['5f3cd303f3759ada3b1da0fe9f63c1f8'[:8],
               '628a844c9d3067033765c0abf4ada356'[:8],
               '5cfd260b132c194f890e386954f82cd5'[:8],
               'e430dea7c7db35fdb16e50397ecfbeec'[:8],
               '660b79253c6c1343678e00e0da560e07'[:8],
               'c077161d06009e072083cddd9f95c9d1'[:8]
               ])

mi = mi.select(picklist=picklist)
for sig, loc in mi.signatures_with_location():
    print(loc, repr(sig))
```
finds all the picklist values across all the files, and does so quite quickly.

(all the other selectors and picklist types will also work just as fast)

**Comment by ctb** [comment]
Ready for review @luizirber @bluegenes

File: src/sourmash/index.py [hunk 1]
```diff
@@ -663,72 +763,101 @@ def consume(self, intersect_mh):
...
+        for row in self.manifest.rows:
+            yield row['signature']
 
     def signatures_with_location(self):
-        for idx, loc in zip(self.index_list, self.source_list):
-            for ss in idx.signatures():
-                yield ss, loc
+        "Return an iterator of tuples (ss, location, internal_location)."
+        for row in self.manifest.rows:
+            yield row['signature'], row['internal_location']
```
Comment (bluegenes) [inline src/sourmash/index.py]: > "Return an iterator of tuples (ss, location, internal_location)."

is `location` part of `row['internal_location']? Or does this docstring need updating?

**Review by bluegenes — APPROVED** [review]
looking good, really excited for this!

File: src/sourmash/index.py (hunk 1 above)
Comment (ctb) [reply src/sourmash/index.py]: good catch, updating :)

Comment (ctb) [reply src/sourmash/index.py]: fixed in c6a8ad74

**Comment by ctb** [comment]
two things I'm thinking of maybe adding, either to this or a new PR -

- [x] look into setting the default file permissions on `SOURMASH-MANIFEST.csv` so that it's a+r; see [this link](https://stackoverflow.com/questions/434641/how-do-i-set-permissions-attributes-on-a-file-in-a-zip-file-using-pythons-zip).
- [ ] the zipfile is opened in uncompressed mode (`zipfile.ZIP_STORED`) and then the signatures are stored in it in compressed format (using `sourmash.signature.save_signatures(..., compression=1)`). This means that the manifest file is _also_ stored in uncompressed mode, because we don't compress it before saving it in the zip file. I'm thinking we might support manifests with .gz filenames, either in addition to or in place of the current `SOURMASH-MANIFEST.csv` filename in zipfile collections.

update: added in https://github.com/sourmash-bio/sourmash/pull/1633

**Comment by bluegenes** [comment]
am I right in assuming that with manifests, we can start doing terrible things like running a gather against all available databases (rather than only using databases with the right ksize/scaled) without measurably increasing search time, because only relevant databases/sigs within databases will actually be loaded?

**Comment by ctb** [comment]
On Wed, Jun 23, 2021 at 04:21:34PM -0700, Tessa Pierce Ward wrote:
> am I right in assuming that with manifests, we can start doing terrible things like running a gather against all available databases (rather than only using databases with the right ksize/scaled) without measurably increasing search time, because only relevant databases/sigs within databases will actually be loaded?

indeed you are correct :)

**Review by luizirber — APPROVED** [review]

**Comment by ctb** [comment]
🎉

**Pull request merged**